src/lib.rs

```diff
@@ -59,368 +82,239 @@ impl<E> From<E> for Error<E> {
-    /// Default values after Power On Reset are: (false, false, LOW)
-    pub fn setup_interrupts(
-        &mut self,
-        mirroring: bool,
-        open_drain: bool,
-        polarity: Polarity,
-    ) -> Result<(), E> {
-        // configure port A
-        self.setup_interrupt_port(Register::IOCONA, mirroring, open_drain, polarity)?;
-
-        // configure port B
-        self.setup_interrupt_port(Register::IOCONB, mirroring, open_drain, polarity)
-    }
+/// MCP23008 Register mapping
+/// See [the datasheet](https://ww1.microchip.com/downloads/en/DeviceDoc/MCP23008-MCP23S08-Data-Sheet-20001919F.pdf) for more
+/// information on the device.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub struct Mcp23008;
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, IntoPrimitive, Default)]
+#[repr(usize)]
+pub enum Mcp23008Pin {
+    #[default]
+    P0 = 0,
+    P1 = 1,
+    P2 = 2,
+    P3 = 3,
+    P4 = 4,
+    P5 = 5,
+    P6 = 6,
+    P7 = 7,
+}
 
-    fn setup_interrupt_port(
-        &mut self,
-        register: Register,
-        mirroring: bool,
-        open_drain: bool,
-        polarity: Polarity,
-    ) -> Result<(), E> {
-        let mut io_conf_value = self.read_register(register)?;
-        io_conf_value = write_bit(io_conf_value, 6, mirroring);
-        io_conf_value = write_bit(io_conf_value, 2, open_drain);
-        io_conf_value = write_bit(io_conf_value, 1, polarity.bit_value());
-        self.write_register(register, io_conf_value)
+impl Map for Mcp23008 {
+    type Pin = Mcp23008Pin;
+    fn map(reg: Register, pin: Self::Pin) -> (u8, usize) {
+        (reg as u8, pin as usize)
     }
+}
 
-    /// Sets up a pin for interrupt.
-    /// Note that the interrupt condition finishes when you read the information about
-    /// the port / value that caused the interrupt or you read the port itself.
-    pub fn setup_interrupt_pin(&mut self, pin: u8, int_mode: InterruptMode) -> Result<(), E> {
-        // set the pin interrupt control (0 means change, 1 means compare against given value)
-        self.update_register_bit(
-            pin,
-            int_mode != InterruptMode::CHANGE,
-            Register::INTCONA,
-            Register::INTCONB,
-        )?;
-
-        // in a RISING interrupt the default value is 0, interrupt is triggered when the pin goes to 1
-        // in a FALLING interrupt the default value is 1, interrupt is triggered when pin goes to 0
-        self.update_register_bit(
-            pin,
-            int_mode == InterruptMode::FALLING,
-            Register::DEFVALA,
-            Register::DEFVALB,
-        )?;
-
-        // enable the pin for interrupt
-        self.update_register_bit(pin, HIGH, Register::GPINTENA, Register::GPINTENB)
-    }
+/// MCP23017/MCP23008, a 16/8-Bit I2C I/O Expander with I2C Interface.
+/// Provide the chip variant via `MAP`.
+#[derive(Clone, Copy, Debug)]
+pub struct Mcp230xx<I2C, MAP> {
+    i2c: I2C,
+    address: u8,
+    variant: core::marker::PhantomData<MAP>,
```

Comment:
Huh, I guess I can use the same structure for `Relay` and then map onto one of the sets or relay pins.
Currently I have the all pins as variants of the same enum. 
https://github.com/quartiq/stabilizer/blob/61d7eeaaf9bc3b54ec446f40a868187f1a8eeb05/src/hardware/driver/relay.rs#L62-L68
So I guess you like the PhantomData+MAP Trait variant better?

Reply:
What do you mean by "structure"? The idiom or `struct`?

`PhantomData` is just there to allow `Mcp230xx` to be generic over `MAP` while not having any struct member that has anything to do with `MAP`.
https://doc.rust-lang.org/std/marker/struct.PhantomData.html#unused-type-parameters

If you feel it helps, use the idiom.
If you want to do code flow depending on which type you are generic over, you need something like this (AFAIK):

https://rust.godbolt.org/z/zqoPsb3a6

Reply:
Yes, I meant the idiom..
Thanks for the additional infos.